cluster_test.go

```diff
@@ -46,6 +46,98 @@ func TestClusterClient(t *T) {
 	assert.Implements(t, new(Client), c)
 }
 
+func makeFailedFlagMap(addrs []string) map[string]bool {
+	failedAddrsFlag := make(map[string]bool)
+	for _, addr := range addrs {
+		failedAddrsFlag[addr] = true
+	}
+	return failedAddrsFlag
+}
+
+func TestClusterInitSync(t *T) {
+	scl := newStubCluster(testTopo)
+	serverAddrs := scl.addrs()
+	{
+		c := scl.newCluster()
+		err := c.Sync(true, false, true)
+		assert.NotNil(t, err)
+	}
+	//part of the addresses are unavailable during the initialization
+	//and recover after that, call Sync to test whether it can work
+	{
+		c, err := scl.newInitSyncErrorCluster(serverAddrs,
+			makeFailedFlagMap(serverAddrs[0:len(serverAddrs)/2]),
+			ClusterWithInitSyncSilent(true))
+		require.Nil(t, err)
+		defer c.Close()
+		//TestClusterSync
+		assertClusterState := func() {
+			require.Nil(t, c.Sync())
+			c.l.RLock()
+			defer c.l.RUnlock()
+			assert.Equal(t, c.topo, scl.topo())
+			assert.Len(t, c.pools, len(c.topo))
+			for _, node := range c.topo {
+				assert.Contains(t, c.pools, node.Addr)
+			}
+		}
+		assertClusterState()
+
+		// cluster is unstable af
+		for i := 0; i < 10; i++ {
+			// find a usabel src/dst
+			var srcStub, dstStub *clusterNodeStub
+			for {
+				srcStub = scl.randStub()
+				dstStub = scl.randStub()
+				if srcStub.addr == dstStub.addr {
+					continue
+				} else if slotRanges := srcStub.slotRanges(); len(slotRanges) == 0 {
+					continue
+				}
+				break
+			}
+
+			// move src's first slot range to dst
+			slotRange := srcStub.slotRanges()[0]
+			t.Logf("moving %d:%d from %s to %s", slotRange[0], slotRange[1], srcStub.addr, dstStub.addr)
+			scl.migrateSlotRange(dstStub.addr, slotRange[0], slotRange[1])
+			assertClusterState()
+		}
+	}
+
+	//part of the addresses are unavailable during the initialization
+	//and recover after that, try Set and Get cmd to test whether it can work
+	{
+		c, err := scl.newInitSyncErrorCluster(serverAddrs,
+			makeFailedFlagMap(serverAddrs[len(serverAddrs)/2:]),
```

Comment:
rather than testing that half of the addresses are failing again, it'd be better to test what happens when _all_ addresses are failing (creating the cluster should fail).

Reply:
Thx for the suggestion. It is essential to test what will happen if all connections fail and I will add the test. In addition, this part is to test whether the get and set commands work normally if some nodes fail during the initialization and then recover. I think this test is necessary.

Reply:
Ok, that makes sense, thanks for explaining it.